docs/source/working_with_aiida/caching.rst

```diff
@@ -4,45 +4,90 @@
+Enabling caching
+----------------
 
-When a calculation is cached, a copy of the original calculation is created. This copy will keep the input links of the new calculation. The outputs of the original calculation are also copied, and linked to the new calculation. This allows for the new calculation to be a separate Node in the provenance graph and, critically, preserves the acyclicity of the graph.
+There are numerous reasons why you may need to re-run calculations you’ve already done before.
+Since AiiDA stores the full provenance of each calculation, it can *detect* whether a calculation has been run before
+and reuse its outputs without wasting computational resources.
 
-Caching is also implemented for Data nodes. This is not very useful in practice (yet), but is an easy way to show how the caching mechanism works:
+This feature is not (yet) enabled by default. In order to enable hashing:
 
-.. ipython::
-    :verbatim:
+ 1. Place the following ``cache_config.yml`` file in your ``.aiida`` configuration folder
+    in order to enable caching for your AiiDA profile (here called ``aiida2``):
 
-    In [1]: from __future__ import print_function
+    .. code:: yaml
 
-    In [2]: from aiida.orm import Str
+        aiida2:
+          default: True
 
-    In [3]: n1 = Str('test string')
+ 2. Run ``verdi -p aiida2 rehash`` in order to compute hashes of all nodes in your existing database
 
-    In [4]: n1.store()
-    Out[4]: u'test string'
+From this point onwards, when you submit a new calculation, AiiDA will compute hashes of the type of calculation and its inputs (see :ref:`caching_matches`). If the hash of the calculation is found to be already present in the database, AiiDA will reuse its results without repeating the actual calculation.
 
-    In [5]: n2 = Str('test string')
+.. _fig_caching:
+.. figure:: include/images/caching.png
+  :align: center
+  :height: 350px
 
-    In [6]: n2.store(use_cache=True)
-    Out[6]: u'test string'
+  When resusing a cached calculation, AiiDA will link up to the input nodes as usual, and make a copy of both the calculation node **C** and its outputs **D3**.
 
-    In [7]: print('UUID of n1:', n1.uuid)
-    UUID of n1: 956109e1-4382-4240-a711-2a4f3b522122
+In order to ensure that the provenance graph with and without caching is the same,
+AiiDA creates a *copy* of the orginial calculation node, linked to the original inputs, as well as a to a copy of the original outputs, as shown in :numref:`fig_caching`.
 
-    In [8]: print('n2 is cached from:', n2.get_extra('_aiida_cached_from'))
-    n2 is cached from: 956109e1-4382-4240-a711-2a4f3b522122
+.. note:: Caching is **not** implemented at the WorkChain/workfunction level (see :ref:`caching_limitations` for details).
+
```

Comment:
I think we need to add some clarification here that for the caching of a `CalculationNode` to happen, one does not need to add the identical input nodes. Rather, since the hash of the input nodes is included in the hash of the `CalculationNode` all inputs merely need to have the same content (and therefore the same hash). I fear giving only this example may suggest that a calculation will only be cached if the exact same input nodes are used. The following example shows that is not the case:
```
#!/usr/bin/env runaiida
from aiida import orm
from aiida.engine import calcfunction
from aiida.manage.caching import enable_caching


@calcfunction
def add(x, y):
    return x + y


x_01 = orm.Int(1)
y_01 = orm.Int(2)

x_02 = orm.Int(1)
y_02 = orm.Int(2)

results_01, node_01 = add.run_get_node(x_01, y_01)
print('Running calcfunction')
print('Inputs pks:', [node.pk for node in node_01.get_incoming().all_nodes()])

results_02, node_02 = add.run_get_node(x_01, y_01)
print('Running calcfunction without caching')
print('Inputs pks:', [node.pk for node in node_02.get_incoming().all_nodes()])

with enable_caching():
    results_03, node_03 = add.run_get_node(x_01, y_01)
print('Running calcfunction with caching')
print('Inputs pks:', [node.pk for node in node_03.get_incoming().all_nodes()])
print('Cache source:', node_03.get_cache_source())

with enable_caching():
    results_04, node_04 = add.run_get_node(x_02, y_02)
print('Running calcfunction with caching and different inputs but with identical value')
print('Inputs pks:', [node.pk for node in node_04.get_incoming().all_nodes()])
print('Cache source:', node_04.get_cache_source())
```
will print:
```
Running calcfunction
('Inputs pks:', [2977, 2976])
Running calcfunction without caching
('Inputs pks:', [2977, 2976])
Running calcfunction with caching
('Inputs pks:', [2977, 2976])
('Cache source:', u'420d73d1-8e18-429f-b9d4-a67bde3f72f3')
Running calcfunction with caching and different inputs but with identical value
('Inputs pks:', [2985, 2984])
('Cache source:', u'420d73d1-8e18-429f-b9d4-a67bde3f72f3')
```
As you can see, despite different input nodes being used in the last case, because the content is identical, the calculation is still cached

Reply:
Thanks - I've added a sentence to the figure caption
>  While this sketch shows **C'** reuse the input nodes of **C'**, it is only the *content* of the input nodes (their hash) that matters.

Reply:
@sphuber Thanks, I will rephrase this.
One question: is the node C' actually an exact copy of the node C (except for the UUID) or not (e.g. does it have different modification time etc.)?
I.e. should one even speak of the node C' being a copy of C or should one rather speak of C' as a new calculation that simply reuses the outputs of C?

Reply:
It will not be an exact copy. If you look at the implementation, it will simply overload all relevant properties (label, description, attributes and repository) from the source node to the "clone" and store it, i.e. the `ctime` and `mtime` will be different. In terms of phrasing then I am not sure if there is any difference between "C' being a copy of C" or "C' as a new calculation that simply reuses the outputs of C". In terms of implementation it is a mix of the two, but I don't think users care about implementation. Ultimately it is a copy with the exact same content that define it. The only difficulty then is defining what properties belong to the "content", e.g. the repository does, but the `ctime` and `mtime` do not.

Reply:
Rephrased now.